### ISSUE TITLE
`from_authorized_user_file` always returns invalid credentials

### ISSUE THREAD

**Issue by busunkim96** [googleapis/google-auth-library-python#501]
`from_user_info` currently always sets the token to None. So a credential created from a JSON is always invalid and must be refreshed.

```
>>> from google.oauth2 import credentials
>>> TOKENS = 'tokens.json' # OAuth2 token storage
>>> creds = credentials.Credentials.from_authorized_user_file(TOKENS)
>>> creds.valid
False
>>> creds.expiry
>>> creds.expired
False
```

 https://github.com/googleapis/google-auth-library-python/blob/772dac6a6512230d32cb0dfae65a1a6aa9015049/google/oauth2/credentials.py#L235-L244

If a token is available, `from_user_info` should use it. `info.get('token', None)`

**Comment by alvyjudy** [comment]
I wonder if this was intentional because the access token is usually set to expire within a short period of time? It is likely expired in this context if it were spun up again from a JSON file.

**Comment by busunkim96** [comment]
@alvyjudy I believe the token is valid for ~1 hour. If the token is expired the refresh logic should kick in, so I don't it's harmful to use `token` from the JSON.


https://github.com/googleapis/google-auth-library-python/blob/772dac6a6512230d32cb0dfae65a1a6aa9015049/google/auth/credentials.py#L69-L76

https://github.com/googleapis/google-auth-library-python/blob/772dac6a6512230d32cb0dfae65a1a6aa9015049/google/auth/credentials.py#L106-L125

**Comment by alvyjudy** [comment]
@busunkim96 I feel the way ``self.expired`` and ``self.valid`` work wouldn't allow us to just pass in the token to ``from_authorized_user_info``

https://github.com/googleapis/google-auth-library-python/blob/772dac6a6512230d32cb0dfae65a1a6aa9015049/google/auth/credentials.py#L53-L67

Since ``self.expiry`` won't be set when a credential is freshly constructed, ``self.expired`` will always return False.

This will cause ``self.valid`` to always return True if ``token`` is passed in

https://github.com/googleapis/google-auth-library-python/blob/772dac6a6512230d32cb0dfae65a1a6aa9015049/google/auth/credentials.py#L69-L76

and ``self.refresh`` won't be called even if the token did expire. Proceeding with the code will result in a ``invalid token`` response when the token is actually valid but expired.

Please correct me if I had made a mistake interpreting the code (I'm quite new here:)

**Comment by alvyjudy** [comment]
A simpler fix would be just to add a line of code to refresh the credential before returning it. I can open a PR if that's desired

**Comment by wescpy** [comment]
I feel that will change our "contract" with the developer. When there's a valid access token stored, users will expect no refresh until after the 3600s have passed for it. It'd be better if we take that access token verbatim and ensure all the flags are set appropriately for however much longer that token's TTL is good for.

**Comment by alvyjudy** [comment]
@wescpy I agree with what you said. It's just that currently ``to_json`` method would strip away the expiry information making it hard to keep track of the TTL. Would it be a better idea  to pickle that information in ``to_json`` and parse it too when picking up the token in ``from_authorized_user_info``?

**Comment by wescpy** [comment]
That sounds much better. Was there a particular reason why it (`expiry`) wasn't included when 1st implemented (to better understand the design decision at the time... wasn't a tz issue was it)?

**Comment by busunkim96** [comment]
@wescpy I don't think so. I believe it was modeled closely after oauth2client, so we might have to take another look there to see if there was a reason for exclusion.

**Comment by wescpy** [comment]
SG. (Of course) `oauth2client` didn't have any issues w/being able to process both active access tokens or run the refresh process if needed. Since it managed the tokens and everyone used that code, it was probably fairly "hardened" (hence why I think at some point in the future, we should also add that feature to this library or `google_auth_oauthlib` [as well as the client libs for the other languages]... it's much better UX if developers didn't have to worry about token timeout issues which are orthogonal to the solutions they're building).

**Comment by dd-ssc** [comment]
I stumbled across this while working my way from the [GMail Python Quickstart](https://developers.google.com/gmail/api/quickstart/python#step_3_set_up_the_sample) towards a production app. That sample app saves/loads the token to/from a local `token.pickle` binary Python pickle file, including the access token.

In our projects, we store all secrets in a [HashiCorp Vault](https://www.vaultproject.io) instance, so a non-binary, text representation is needed for persistence. As a first step towards that, I replaced the local pickle file by a local json file.

When I run the app subsquently during development, the access token is still perfectly valid; there is no reason to refresh. Still, the credentials returned by `from_authorized_user_info()` are invalid.
As a result, it isn't even attempted to refresh the access token by means of the refresh token - the entire user auth process is started _every time_: Browser window opens, user logs into their GMail / GSuite account, etc.

I think I can work around this for the time being by just re-adding the access token to the credentials, but from my point of view, this just feels like a bug.

**Comment by dd-ssc** [comment]
Update: The workaround for this issue is trivial, posting it here in case anyone else runs into it:

```
import json
from google.oauth2.credentials import Credentials

def load_user_secrets_from_local(user_secrets_file):
    if exists(user_secrets_file):
        with open(user_secrets_file, 'r') as stream:
            creds_json = json.load(stream)
        creds = Credentials.from_authorized_user_info(creds_json)
        # workaround for
        # https://github.com/googleapis/google-auth-library-python/issues/501
        creds.token = creds_json['token']
        return creds
    return None
```

**Comment by wescpy** [comment]
A workaround is good but not a permanent fix for this bug, so I spent the day working on it and submitted for review as </pull/589>. Thanks to @dd-ssc for the inspiration!

**Comment by wescpy** [comment]
While the PR has been merged, this isn't officially "fixed" until the next release. However, those who are being bitten by this can at least grab the latest [google/oauth2/credentials.py](https://github.com/googleapis/google-auth-library-python/blob/master/google/oauth2/credentials.py) and replace their faulty one.

**Comment by busunkim96** [comment]
Closed by #589
